File: payment-core/go-services/internal/messaging/transactional_outbox.go
package messaging

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"sync"
	"time"

	"github.com/google/uuid"
)

type OutboxStatus string

const (
	OutboxStatusPending    OutboxStatus = "pending"
	OutboxStatusProcessing OutboxStatus = "processing"
	OutboxStatusSent       OutboxStatus = "sent"
	OutboxStatusFailed     OutboxStatus = "failed"
	OutboxStatusDeadLetter OutboxStatus = "dead_letter"
)

type OutboxMessage struct {
	ID            string                 `json:"id"`
	AggregateType string                 `json:"aggregate_type"`
	AggregateID   string                 `json:"aggregate_id"`
	EventType     string                 `json:"event_type"`
	Payload       map[string]interface{} `json:"payload"`
	Destination   string                 `json:"destination"`
	Status        OutboxStatus           `json:"status"`
	RetryCount    int                    `json:"retry_count"`
	MaxRetries    int                    `json:"max_retries"`
	CreatedAt     time.Time              `json:"created_at"`
	ProcessedAt   *time.Time             `json:"processed_at,omitempty"`
	LastError     string                 `json:"last_error,omitempty"`
	Metadata      map[string]interface{} `json:"metadata,omitempty"`
}

type InboxMessage struct {
	ID             string    `json:"id"`
	MessageID      string    `json:"message_id"`
	EventType      string    `json:"event_type"`
	ConsumerID     string    `json:"consumer_id"`
	ProcessedAt    time.Time `json:"processed_at"`
	IdempotencyKey string    `json:"idempotency_key"`
}

type OutboxConfig struct {
	PollIntervalMs        int
	BatchSize             int
	MaxRetries            int
	RetryDelayMs          int
	DeadLetterAfterRetries int
}

type Publisher func(ctx context.Context, message *OutboxMessage) error

type TransactionalOutbox struct {
	mu              sync.RWMutex
	config          OutboxConfig
	outboxStore     map[string]*OutboxMessage
	deadLetterQueue []*OutboxMessage
	publishers      map[string]Publisher
	isRunning       bool
	stopChan        chan struct{}
	eventHandlers   map[string][]func(*OutboxMessage)
}

func NewTransactionalOutbox(config *OutboxConfig) *TransactionalOutbox {
	cfg := OutboxConfig{
		PollIntervalMs:        1000,
		BatchSize:             100,
		MaxRetries:            5,
		RetryDelayMs:          5000,
		DeadLetterAfterRetries: 10,
	}

	if config != nil {
		if config.PollIntervalMs > 0 {
			cfg.PollIntervalMs = config.PollIntervalMs
		}
		if config.BatchSize > 0 {
			cfg.BatchSize = config.BatchSize
		}
		if config.MaxRetries > 0 {
			cfg.MaxRetries = config.MaxRetries
		}
		if config.RetryDelayMs > 0 {
			cfg.RetryDelayMs = config.RetryDelayMs
		}
		if config.DeadLetterAfterRetries > 0 {
			cfg.DeadLetterAfterRetries = config.DeadLetterAfterRetries
		}
	}

	return &TransactionalOutbox{
		config:          cfg,
		outboxStore:     make(map[string]*OutboxMessage),
		deadLetterQueue: make([]*OutboxMessage, 0),
		publishers:      make(map[string]Publisher),
		eventHandlers:   make(map[string][]func(*OutboxMessage)),
		stopChan:        make(chan struct{}),
	}
}

func (to *TransactionalOutbox) RegisterPublisher(destination string, publisher Publisher) {
	to.mu.Lock()
	defer to.mu.Unlock()
	to.publishers[destination] = publisher
}

func (to *TransactionalOutbox) On(event string, handler func(*OutboxMessage)) {
	to.mu.Lock()
	defer to.mu.Unlock()
	to.eventHandlers[event] = append(to.eventHandlers[event], handler)
}

func (to *TransactionalOutbox) emit(event string, message *OutboxMessage) {
	to.mu.RLock()
	handlers := to.eventHandlers[event]
	to.mu.RUnlock()

	for _, handler := range handlers {
		go handler(message)
	}
}

type AddMessageParams struct {
	AggregateType string
	AggregateID   string
	EventType     string
	Payload       map[string]interface{}
	Destination   string
	Metadata      map[string]interface{}
}

func (to *TransactionalOutbox) AddMessage(params AddMessageParams) (*OutboxMessage, error) {
	to.mu.Lock()
	defer to.mu.Unlock()

	message := &OutboxMessage{
		ID:            uuid.New().String(),
		AggregateType: params.AggregateType,
		AggregateID:   params.AggregateID,
		EventType:     params.EventType,
		Payload:       params.Payload,
		Destination:   params.Destination,
		Status:        OutboxStatusPending,
		RetryCount:    0,
		MaxRetries:    to.config.MaxRetries,
		CreatedAt:     time.Now(),
		Metadata:      params.Metadata,
	}

	to.outboxStore[message.ID] = message
	to.emit("messageAdded", message)
	return message, nil
}

func (to *TransactionalOutbox) ProcessMessages(ctx context.Context) (int, error) {
	to.mu.Lock()

	pendingMessages := make([]*OutboxMessage, 0)
	for _, m := range to.outboxStore {
		if (m.Status == OutboxStatusPending || m.Status == OutboxStatusFailed) &&
			m.RetryCount < to.config.DeadLetterAfterRetries {
			pendingMessages = append(pendingMessages, m)
		}
		if len(pendingMessages) >= to.config.BatchSize {
			break
		}
	}
	to.mu.Unlock()

	processed := 0

	for _, message := range pendingMessages {
		to.mu.Lock()
		message.Status = OutboxStatusProcessing
		to.mu.Unlock()

		to.mu.RLock()
		publisher, ok := to.publishers[message.Destination]
		to.mu.RUnlock()

		if !ok {
			to.mu.Lock()
			message.RetryCount++
			message.LastError = fmt.Sprintf("no publisher registered for destination: %s", message.Destination)
			message.Status = OutboxStatusFailed
			to.mu.Unlock()
			to.emit("messageFailed", message)
			continue
		}

		err := publisher(ctx, message)
		if err != nil {
			to.mu.Lock()
			message.RetryCount++
			message.LastError = err.Error()

			if message.RetryCount >= to.config.DeadLetterAfterRetries {
				message.Status = OutboxStatusDeadLetter
				to.deadLetterQueue = append(to.deadLetterQueue, message)
				delete(to.outboxStore, message.ID)
				to.mu.Unlock()
				to.emit("messageDeadLettered", message)
			} else {
				message.Status = OutboxStatusFailed
				to.mu.Unlock()
				to.emit("messageFailed", message)
			}
			continue
		}

		to.mu.Lock()
		now := time.Now()
		message.Status = OutboxStatusSent
		message.ProcessedAt = &now
		to.mu.Unlock()

		processed++
		to.emit("messageSent", message)
	}

	return processed, nil
}

func (to *TransactionalOutbox) Start(ctx context.Context) {
	to.mu.Lock()
	if to.isRunning {
		to.mu.Unlock()
		return
	}
	to.isRunning = true
	to.stopChan = make(chan struct{})
	to.mu.Unlock()

	ticker := time.NewTicker(time.Duration(to.config.PollIntervalMs) * time.Millisecond)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return
		case <-to.stopChan:
			return
		case <-ticker.C:
			to.ProcessMessages(ctx)
		}
	}
}

func (to *TransactionalOutbox) Stop() {
	to.mu.Lock()
	defer to.mu.Unlock()

	if !to.isRunning {
		return
	}

	to.isRunning = false
	close(to.stopChan)
}

type OutboxStats struct {
	Pending    int `json:"pending"`
	Processing int `json:"processing"`
	Sent       int `json:"sent"`
	Failed     int `json:"failed"`
	DeadLetter int `json:"dead_letter"`
}

func (to *TransactionalOutbox) GetStats() OutboxStats {
	to.mu.RLock()
	defer to.mu.RUnlock()

	stats := OutboxStats{}
	for _, m := range to.outboxStore {
		switch m.Status {
		case OutboxStatusPending:
			stats.Pending++
		case OutboxStatusProcessing:
			stats.Processing++
		case OutboxStatusSent:
			stats.Sent++
		case OutboxStatusFailed:
			stats.Failed++
		}
	}
	stats.DeadLetter = len(to.deadLetterQueue)
	return stats
}

func (to *TransactionalOutbox) GetDeadLetterQueue() []*OutboxMessage {
	to.mu.RLock()
	defer to.mu.RUnlock()

	result := make([]*OutboxMessage, len(to.deadLetterQueue))
	copy(result, to.deadLetterQueue)
	return result
}

func (to *TransactionalOutbox) RetryDeadLetter(messageID string) error {
	to.mu.Lock()
	defer to.mu.Unlock()

	for i, m := range to.deadLetterQueue {
		if m.ID == messageID {
			to.deadLetterQueue = append(to.deadLetterQueue[:i], to.deadLetterQueue[i+1:]...)
			m.Status = OutboxStatusPending
			m.RetryCount = 0
			m.LastError = ""
			to.outboxStore[m.ID] = m
			to.emit("deadLetterRetried", m)
			return nil
		}
	}

	return errors.New("message not found in dead letter queue")
}

type ConsumerDeduplication struct {
	mu         sync.RWMutex
	consumerID string
	inboxStore map[string]*InboxMessage
}

func NewConsumerDeduplication(consumerID string) *ConsumerDeduplication {
	return &ConsumerDeduplication{
		consumerID: consumerID,
		inboxStore: make(map[string]*InboxMessage),
	}
}

func (cd *ConsumerDeduplication) getDedupeKey(messageID, idempotencyKey string) string {
	key := idempotencyKey
	if key == "" {
		key = messageID
	}
	return fmt.Sprintf("%s:%s", cd.consumerID, key)
}

func (cd *ConsumerDeduplication) IsDuplicate(messageID, idempotencyKey string) bool {
	cd.mu.RLock()
	defer cd.mu.RUnlock()

	key := cd.getDedupeKey(messageID, idempotencyKey)
	_, exists := cd.inboxStore[key]
	return exists
}

func (cd *ConsumerDeduplication) MarkProcessed(messageID, eventType, idempotencyKey string) {
	cd.mu.Lock()
	defer cd.mu.Unlock()

	key := cd.getDedupeKey(messageID, idempotencyKey)
	iKey := idempotencyKey
	if iKey == "" {
		iKey = messageID
	}

	cd.inboxStore[key] = &InboxMessage{
		ID:             uuid.New().String(),
		MessageID:      messageID,
		EventType:      eventType,
		ConsumerID:     cd.consumerID,
		ProcessedAt:    time.Now(),
		IdempotencyKey: iKey,
	}
}

type ProcessResult struct {
	Processed bool
	Result    interface{}
	Duplicate bool
}

func (cd *ConsumerDeduplication) ProcessWithDeduplication(
	messageID, eventType, idempotencyKey string,
	processor func() (interface{}, error),
) (*ProcessResult, error) {
	if cd.IsDuplicate(messageID, idempotencyKey) {
		return &ProcessResult{Processed: false, Duplicate: true}, nil
	}

	result, err := processor()
	if err != nil {
		return nil, err
	}

	cd.MarkProcessed(messageID, eventType, idempotencyKey)
	return &ProcessResult{Processed: true, Result: result, Duplicate: false}, nil
}

type KafkaMessage struct {
	Topic   string
	Key     string
	Value   []byte
	Headers map[string]string
}

func CreateKafkaPublisher(sendFunc func(ctx context.Context, msg *KafkaMessage) error) Publisher {
	return func(ctx context.Context, message *OutboxMessage) error {
		payload := map[string]interface{}{
			"id":            message.ID,
			"type":          message.EventType,
			"aggregate_type": message.AggregateType,
			"aggregate_id":  message.AggregateID,
			"payload":       message.Payload,
			"timestamp":     message.CreatedAt.Format(time.RFC3339),
			"metadata":      message.Metadata,
		}

		value, err := json.Marshal(payload)
		if err != nil {
			return fmt.Errorf("failed to marshal message: %w", err)
		}

		kafkaMsg := &KafkaMessage{
			Topic: message.Destination,
			Key:   message.AggregateID,
			Value: value,
			Headers: map[string]string{
				"event-type":     message.EventType,
				"aggregate-type": message.AggregateType,
				"message-id":     message.ID,
			},
		}

		return sendFunc(ctx, kafkaMsg)
	}
}

File: server/messaging/transactionalOutbox.ts
import crypto from 'crypto';
import { EventEmitter } from 'events';

export interface OutboxMessage {
  id: string;
  aggregateType: string;
  aggregateId: string;
  eventType: string;
  payload: Record<string, any>;
  destination: string;
  status: OutboxStatus;
  retryCount: number;
  maxRetries: number;
  createdAt: Date;
  processedAt?: Date;
  lastError?: string;
  metadata?: Record<string, any>;
}

export type OutboxStatus = 'pending' | 'processing' | 'sent' | 'failed' | 'dead_letter';

export interface InboxMessage {
  id: string;
  messageId: string;
  eventType: string;
  consumerId: string;
  processedAt: Date;
  idempotencyKey: string;
}

export interface OutboxConfig {
  pollIntervalMs: number;
  batchSize: number;
  maxRetries: number;
  retryDelayMs: number;
  deadLetterAfterRetries: number;
}

const outboxStore = new Map<string, OutboxMessage>();
const inboxStore = new Map<string, InboxMessage>();
const deadLetterQueue: OutboxMessage[] = [];

export class TransactionalOutbox extends EventEmitter {
  private config: OutboxConfig;
  private isRunning: boolean = false;
  private pollInterval: NodeJS.Timeout | null = null;
  private publishers: Map<string, (message: OutboxMessage) => Promise<void>> = new Map();

  constructor(config?: Partial<OutboxConfig>) {
    super();
    this.config = {
      pollIntervalMs: config?.pollIntervalMs || 1000,
      batchSize: config?.batchSize || 100,
      maxRetries: config?.maxRetries || 5,
      retryDelayMs: config?.retryDelayMs || 5000,
      deadLetterAfterRetries: config?.deadLetterAfterRetries || 10
    };
  }

  registerPublisher(destination: string, publisher: (message: OutboxMessage) => Promise<void>): void {
    this.publishers.set(destination, publisher);
  }

  async addMessage(params: {
    aggregateType: string;
    aggregateId: string;
    eventType: string;
    payload: Record<string, any>;
    destination: string;
    metadata?: Record<string, any>;
  }): Promise<OutboxMessage> {
    const message: OutboxMessage = {
      id: crypto.randomUUID(),
      aggregateType: params.aggregateType,
      aggregateId: params.aggregateId,
      eventType: params.eventType,
      payload: params.payload,
      destination: params.destination,
      status: 'pending',
      retryCount: 0,
      maxRetries: this.config.maxRetries,
      createdAt: new Date(),
      metadata: params.metadata
    };

    outboxStore.set(message.id, message);
    this.emit('messageAdded', message);
    return message;
  }

  async processMessages(): Promise<number> {
    const pendingMessages = Array.from(outboxStore.values())
      .filter(m => m.status === 'pending' || m.status === 'failed')
      .filter(m => m.retryCount < this.config.deadLetterAfterRetries)
      .slice(0, this.config.batchSize);

    let processed = 0;

    for (const message of pendingMessages) {
      try {
        message.status = 'processing';
        
        const publisher = this.publishers.get(message.destination);
        if (!publisher) {
          throw new Error(`No publisher registered for destination: ${message.destination}`);
        }

        await publisher(message);
        
        message.status = 'sent';
        message.processedAt = new Date();
        processed++;
        
        this.emit('messageSent', message);
      } catch (error) {
        message.retryCount++;
        message.lastError = (error as Error).message;
        
        if (message.retryCount >= this.config.deadLetterAfterRetries) {
          message.status = 'dead_letter';
          deadLetterQueue.push(message);
          outboxStore.delete(message.id);
          this.emit('messageDeadLettered', message);
        } else {
          message.status = 'failed';
          this.emit('messageFailed', message, error);
        }
      }
    }

    return processed;
  }

  start(): void {
    if (this.isRunning) return;
    
    this.isRunning = true;
    this.pollInterval = setInterval(async () => {
      try {
        await this.processMessages();
      } catch (error) {
        this.emit('error', error);
      }
    }, this.config.pollIntervalMs);
    
    this.emit('started');
  }

  stop(): void {
    if (!this.isRunning) return;
    
    this.isRunning = false;
    if (this.pollInterval) {
      clearInterval(this.pollInterval);
      this.pollInterval = null;
    }
    
    this.emit('stopped');
  }

  getStats(): {
    pending: number;
    processing: number;
    sent: number;
    failed: number;
    deadLetter: number;
  } {
    const messages = Array.from(outboxStore.values());
    return {
      pending: messages.filter(m => m.status === 'pending').length,
      processing: messages.filter(m => m.status === 'processing').length,
      sent: messages.filter(m => m.status === 'sent').length,
      failed: messages.filter(m => m.status === 'failed').length,
      deadLetter: deadLetterQueue.length
    };
  }

  getDeadLetterQueue(): OutboxMessage[] {
    return [...deadLetterQueue];
  }

  async retryDeadLetter(messageId: string): Promise<boolean> {
    const index = deadLetterQueue.findIndex(m => m.id === messageId);
    if (index === -1) return false;

    const message = deadLetterQueue.splice(index, 1)[0];
    message.status = 'pending';
    message.retryCount = 0;
    message.lastError = undefined;
    
    outboxStore.set(message.id, message);
    this.emit('deadLetterRetried', message);
    return true;
  }
}

export class ConsumerDeduplication {
  private consumerId: string;

  constructor(consumerId: string) {
    this.consumerId = consumerId;
  }

  async isDuplicate(messageId: string, idempotencyKey?: string): Promise<boolean> {
    const key = this.getDedupeKey(messageId, idempotencyKey);
    return inboxStore.has(key);
  }

  async markProcessed(messageId: string, eventType: string, idempotencyKey?: string): Promise<void> {
    const key = this.getDedupeKey(messageId, idempotencyKey);
    
    const record: InboxMessage = {
      id: crypto.randomUUID(),
      messageId,
      eventType,
      consumerId: this.consumerId,
      processedAt: new Date(),
      idempotencyKey: idempotencyKey || messageId
    };

    inboxStore.set(key, record);
  }

  async processWithDeduplication<T>(
    messageId: string,
    eventType: string,
    processor: () => Promise<T>,
    idempotencyKey?: string
  ): Promise<{ processed: boolean; result?: T; duplicate: boolean }> {
    if (await this.isDuplicate(messageId, idempotencyKey)) {
      return { processed: false, duplicate: true };
    }

    const result = await processor();
    await this.markProcessed(messageId, eventType, idempotencyKey);
    
    return { processed: true, result, duplicate: false };
  }

  private getDedupeKey(messageId: string, idempotencyKey?: string): string {
    return `${this.consumerId}:${idempotencyKey || messageId}`;
  }
}

export function createKafkaPublisher(kafkaProducer: any): (message: OutboxMessage) => Promise<void> {
  return async (message: OutboxMessage) => {
    await kafkaProducer.send({
      topic: message.destination,
      messages: [{
        key: message.aggregateId,
        value: JSON.stringify({
          id: message.id,
          type: message.eventType,
          aggregateType: message.aggregateType,
          aggregateId: message.aggregateId,
          payload: message.payload,
          timestamp: message.createdAt.toISOString(),
          metadata: message.metadata
        }),
        headers: {
          'event-type': message.eventType,
          'aggregate-type': message.aggregateType,
          'message-id': message.id
        }
      }]
    });
  };
}

let outboxInstance: TransactionalOutbox | null = null;

export function getTransactionalOutbox(): TransactionalOutbox {
  if (!outboxInstance) {
    outboxInstance = new TransactionalOutbox();
  }
  return outboxInstance;
}

export default TransactionalOutbox;
